### ISSUE TITLE
vmapping list appends and performance

### ISSUE THREAD

**Issue by choltz95** [google/jax#3297]
Hi, I really love this library - I just wish I knew how to use it properly. 

I'm sorry if this is not the right place for this. I asked this question on stackoverflow, but never got a response. 

I am working with Jax and Stax. I have added an alternative feedforward function to my stax network and I am vmaping over samples in each batch. It works well. The issue is it's very slow (~20x vmapped compared to vanilla feedforward not vmapped), and I am not sure why. Does anything look incorrect from a cursory glance? I have verified that the fn is compiling (by printing variables -> Traced<ShapedArray(...))
```
def pwl_fun(params, configs, **kwargs):
    lambdas = [np.diag(config) for config in configs]
    js = [np.diag(-2 * config + 1) for config in configs]

    # Compute Z_k = W_k * x + b_k for each layer
    wks = [params[0][0].T] #
    bks = [params[0][1].T]
    i = 0

    for p in params[1:]: # 2
        current_wk = wks[-1]
        current_bk = bks[-1]
        current_lambda = lambdas[i]
        c_len_p = (len(p) == 2)
        i = lax.cond(c_len_p, i, lambda i: i+1, i, lambda i: i)

        if c_len_p:
            precompute = np.matmul(p[0].T,current_lambda)
            wks.append(np.matmul(precompute,current_wk))
            bks.append(np.matmul(precompute,current_bk) + p[1])

    a_stack = []
    b_stack = []        
    for j, wk, bk in zip(js, wks, bks):
        a_stack.append(np.matmul(j,wk))
        b_stack.append(-np.matmul(j,bk))

    polytope_A = np.concatenate(a_stack)
    polytope_b = np.concatenate(b_stack)

    return polytope_A, polytope_b, wks[-1], bks[-1]
```
I have also modified the above fn to remove instances of control flow or change them into lax.conds (e.g. that "if c_len_p" just determines if the current param is a weight matrix), but it does not help.  

I know this style is not properly functional. Unfortunately I have to use lists here for some of the variables (e.g. elements of config don't have the same dimension). Could that be the problem? Would passing the batched data as a param instead and shifting vmaps inside this function do anything? I'm a bit lost & would definitely appreciate advice. 

For reference,
here is the associated pytorch (basically identical to the above): https://github.com/revbucket/geometric-certificates/blob/master/plnn.py#L133 

Edit 1: Using Jax '0.1.64'. Working with a simplified version without any loops/conditionals/lists. No change in runtime. I feel like it's much slower than it should be:
```
def pwl_fun(params, configs, **kwargs):
        l = np.diag(configs)
        j = np.diag(-2 * configs + 1)
        
        current_wk = params[0][0].T
        current_bk = params[0][1].T
        current_lambda = l

        polytope_A = np.matmul(j,current_wk)
        polytope_b = -np.matmul(j,current_bk)

        precompute = np.matmul(params[2][0].T,current_lambda)
        wks = np.matmul(precompute,current_wk)
        bks = np.matmul(precompute,current_bk) + params[2][1]

        return polytope_A, polytope_b, wks, bks
```

Edit 2: I verified that the performance is bottlenecked by the autograd comp.

**Comment by choltz95** [comment]
FYI I have added colab link: https://colab.research.google.com/drive/1HNpk2ihIgt7CuzqDg9mH22x2m1pnhLIB?usp=sharing

No luck yet & no leads - I guess the slowness may be unavoidable?

**Comment by choltz95** [comment]
Sorry for the spam.

I think I have identified the offending line. 
`j = np.diag(-2 * configs + 1)`

configs is function of the input to the layer: a binarized relu applied to the layer input.

If I instead define configs to just be a constant or Bernoulli rv, there is no slowdown.

I don't have to backprop through j, so I thought that preventing gradients propagating through configs would fix this using lax.stop_gradient, but it doesn't help. 
`j = np.diag(lax.stop_gradient(configs))`
